usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -1021,34 +1032,78 @@ static CK_RV ep11tok_pkey_update(STDLL_TokData_t *tokdata, SESSION *session,
         goto done;
     }
 
-    /* Check if the new pkey's verification pattern matches the one in
-     * ep11_data. This should always be the case, because we just
-     * created the pkey with the current MK. */
+    if (pthread_mutex_lock(&ep11_data->pkey_mutex)) {
+        TRACE_ERROR("%s Failed to lock pkey lock\n", __func__);
+        ret = CKR_CANT_LOCK;
+        goto done;
+    }
+
+    /*
+     * Check if the new pkey's verification pattern matches the one in
+     * ep11_data. This should always be the case, except there was a live
+     * guest relocation (LGR) in the middle of the process.
+     */
     vp_offset = pkey_attr->ulValueLen - PKEY_MK_VP_LENGTH;
     if (memcmp(&ep11_data->pkey_mk_vp, (CK_BYTE *)pkey_attr->pValue + vp_offset,
                PKEY_MK_VP_LENGTH) != 0) {
-        TRACE_ERROR("vp of this pkey does not match with the one in ep11_data (should not occur)\n");
+        TRACE_ERROR("vp of this pkey does not match with the one in ep11_data\n");
         ret = CKR_FUNCTION_FAILED;
-        goto done;
     }
 
     if (aes_xts) {
-        /* Check if the new pkey's verification pattern matches the one in
-         * ep11_data. This should always be the case, because we just
-         * created the pkey with the current MK.
+        /*
+         * Check if the new pkey's verification pattern matches the one in
+         * ep11_data. This should always be the case, except there was a live
+         * guest relocation (LGR) in the middle of the process.
          * AES XTS has two keys, two keys are concatenated.
-         * Second key is checked above and the first key is checked here */
+         * Second key is checked above and the first key is checked here
+         */
         vp_offset = pkey_attr->ulValueLen / 2 - PKEY_MK_VP_LENGTH;
         if (memcmp(&ep11_data->pkey_mk_vp, (CK_BYTE *)pkey_attr->pValue + vp_offset,
                    PKEY_MK_VP_LENGTH) != 0) {
-            TRACE_ERROR("vp of this pkey does not match with the one in ep11_data (should not occur)\n");
+            TRACE_ERROR("vp of this pkey does not match with the one in ep11_data\n");
             ret = CKR_FUNCTION_FAILED;
+        }
+    }
+
+    /*
+     * Recreating the firmware mkvp via ep11tok_pkey_get_firmware_mk_vp below
+     * does only work when pkey_wrap_support_checked is 0. Set it to 0 while
+     * the mutex lock is still obtained.
+     */
+    if (ret != CKR_OK)
+        __sync_and_and_fetch(&ep11_data->pkey_wrap_support_checked, 0);
+
+    if (pthread_mutex_unlock(&ep11_data->pkey_mutex))
+        TRACE_ERROR("%s Failed to unlock pkey lock\n", __func__);
```

Comment:
Please use { .. } (TRACE_ERROR is a macro).

Reply:
fixed.